src/test/java/org/jenkinsci/plugins/workflow/log/LogStorageTestBase.java

```diff
@@ -171,9 +186,66 @@ protected static void close(TaskListener listener) throws Exception {
     }
 
     @Test public void remoting() throws Exception {
+        remotingParameterized(RemotingTestVariant.Regular);
+    }
+
+    /**
+     * Verifies that flushing explicitly makes {@link BufferedBuildListener} and {@link GCFlushedOutputStream} unnecessary.
+     * Only relevant for {@link LogStorage} implementations that use {@link BufferedBuildListener}.
+     */
+    @Test public void remotingNormalFlushOnly() throws Exception {
+        remotingParameterized(RemotingTestVariant.ManualFlushOnly);
+    }
+
+    /**
+     * Verifies the behavior of {@link BufferedBuildListener}.
+     * Only relevant for {@link LogStorage} implementations that use {@link BufferedBuildListener}.
+     */
+    @Test public void remotingDelayedAutoFlushOnly() throws Exception {
+        // To make this test fail, set `DelayBufferedOutputStream.Tuning.DEFAULT.*RecurrencePeriod` to very large values.
+        remotingParameterized(RemotingTestVariant.DelayedAutoFlushOnly);
+    }
+
+    /**
+     * Verifies the behavior of {@link GCFlushedOutputStream}.
+     * Only relevant for {@link LogStorage} implementations that use {@link BufferedBuildListener}.
+     */
+    @Test public void remotingGcFlushOnly() throws Exception {
+        // To make this test fail, set `GCFlushedOutputStream.DISABLED` to true.
+        remotingParameterized(RemotingTestVariant.GcFlushOnly);
+    }
+
+    private enum RemotingTestVariant {
+        Regular(false, false, false),
+        ManualFlushOnly(true, true, false),
+        DelayedAutoFlushOnly(true, false, true),
+        GcFlushOnly(false, true, true);
+
+        boolean disableGcFlush;
+        boolean disableDelayedAutoFlush;
+        boolean disableManualFlush;
+
+        RemotingTestVariant(boolean disableGcFlush, boolean disableDelayedAutoFlush, boolean disableManualFlush) {
+            this.disableGcFlush = disableGcFlush;
+            this.disableDelayedAutoFlush = disableDelayedAutoFlush;
+            this.disableManualFlush = disableManualFlush;
+        }
+    }
+    private void remotingParameterized(RemotingTestVariant variant) throws Exception {
         logging.capture(100).record(Channel.class, Level.WARNING);
+        if (variant.disableDelayedAutoFlush) {
+            DelayBufferedOutputStream.Tuning.DEFAULT.minRecurrencePeriod = TimeUnit.HOURS.toMillis(1);
+            DelayBufferedOutputStream.Tuning.DEFAULT.maxRecurrencePeriod = TimeUnit.HOURS.toMillis(1);
+        }
+        if (variant.disableGcFlush) {
+            GCFlushedOutputStream.DISABLED = true;
+        }
         LogStorage ls = createStorage();
         TaskListener overall = ls.overallListener();
+        if (variant != RemotingTestVariant.Regular) {
+            assumeThat("Skipping BufferedBuildListener-specific tests because listener is " + overall,
+                    overall, instanceOf(BufferedBuildListener.class));
```

Comment:
This is unnecessary, i.e. the new tests pass against `pipeline-cloudwatch-logs`, but it seemed somewhat pointless to run tests that configure `BufferedBuildListener` against unrelated listener implementations.

If desired, we could instead move the new tests down into `FileLogStorageTest` add make the tests here more generic. For example, we'd just have one test that checks that if you write a line and do not flush, the line gets written if you wait for a few seconds. `pipeline-cloudwatch-logs` would still pass this test without waiting because it doesn't buffer more than a single line. `FileLogStorage` would only pass after a few seconds due to `DelayBufferedOutputStream`.

Reply:
> if you write a line and do not flush, the line gets written if you wait for a few seconds

Fine, I guess, but no one should rely on that; remote code printing to a `TaskListener` is expected to flush explicitly.